### PR TITLE
feat: various

### PR DESCRIPTION
**Pull request by zeruth**
adds OpObj,OpLoc,OpNpc,OpPlayer decoders
handle IfButton decoding properly
adds packet g2_alt1/2 
adds packet g4_alt1/2/3

(Read Login Packet)https://github.com/LostCityRS/Engine/pull/5/commits/2a22c3eb051ef4d4b3083a91d0ae1f2878beedc2
(Fix Music Tab)https://github.com/LostCityRS/Engine/pull/5/commits/86a60a7306dcd6c3809ce06318895bc1051480ab

### CODE REVIEW

File: src/network/os1/client/handler/game/LocExamineHandler.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import MessageHandler from '#/network/client/handler/MessageHandler.ts';
+import MessageGame from "#/network/server/model/game/MessageGame.js";
+import type LocExamine from "#/network/client/model/game/LocExamine.js";
+
+export default class LocExamineHandler extends MessageHandler {
+    handle(message: LocExamine, player: NetworkPlayer): boolean {
+
+        //todo: remove easter egg
+        if (message.id === 1530) {
+            player.write(new MessageGame('Who\'s there?'));
```
Comment (Pazaz) [inline src/network/os1/client/handler/game/LocExamineHandler.ts]: can use double-quotes to single-quote inside :)
"Who's there?"

File: src/network/os1/client/prot/game/GameClientRepository.ts [hunk 2]
```diff
@@ -30,5 +82,31 @@ export default class GameClientRepository extends ClientRepository {
         this.bind(new ClientCheatDecoder(), new ClientCheatHandler())
         this.bind(new IfButtonDecoder(), new IfButtonHandler())
         this.bind(new ClanDecoder(), new ClanHandler())
+        this.bind(new OpPlayer1Decoder(), new OpPlayer1Handler())
```
Comment (Pazaz) [inline src/network/os1/client/prot/game/GameClientRepository.ts]: I would make all of these into single decoder/handlers per *type* (OpPlayer, OpNpc, OpLoc, OpPlayerU, OpNpcU, OpPlayerT, ...), you can pass the opcode into the decoder and track "option" / "op" in the message.

I can give an example on how to implement if you need. Your message model becomes:
```ts
class OpPlayer {
    op: number;
    idx: number;
}
```
And you use that op in the handler.

Comment (zeruth) [reply src/network/os1/client/prot/game/GameClientRepository.ts]: I think I got it the way you wanted. cleaned up the PR quite a bit so thank you

File: src/network/os1/client/prot/game/GameClientRepository.ts [hunk 3]
```diff
@@ -30,5 +63,31 @@ export default class GameClientRepository extends ClientRepository {
         this.bind(new ClientCheatDecoder(), new ClientCheatHandler())
         this.bind(new IfButtonDecoder(), new IfButtonHandler())
         this.bind(new ClanDecoder(), new ClanHandler())
+        this.bind(new OpPlayer1Decoder(), new OpPlayerHandler())
```
Comment (Pazaz) [inline src/network/os1/client/prot/game/GameClientRepository.ts]: Much better. You could do the same for the decoders to reduce some boilerplate too! Pass op and opcode into "OpPlayerDecoder":

```ts
this.bind(new OpPlayerDecoder(1, 245), new OpPlayerHandler());
this.bind(new OpPlayerDecoder(2, 146), new OpPlayerHandler());
this.bind(new OpPlayerDecoder(3, 102), new OpPlayerHandler());
this.bind(new OpPlayerDecoder(4, 78), new OpPlayerHandler());
```

```ts
export default class OpPlayerDecoder extends MessageDecoder {
    size = 2;
    limit = GameClientLimit.CLIENT;

    constructor(op: number, opcode: number) {
        this.op = op;
        this.opcode = opcode;
    }

    read(buf: Packet): ClientMessage {
        let playerId = -1;
        if (this.op === 1) {
            playerId = buf.g2();
        } else if (this.op === 2) {
            playerId = buf.g2();
        } else if (this.op === 3) {
            playerId = buf.g2_alt1();
        }

        return new OpPlayer(this.op, playerId);
    }
}
```

File: src/network/os1/client/prot/game/GameClientRepository.ts [hunk 4]
```diff
@@ -30,5 +63,31 @@ export default class GameClientRepository extends ClientRepository {
...
+        this.bind(new OpLoc2Decoder(), new OpLocHandler())
+        this.bind(new OpLoc3Decoder(), new OpLocHandler())
+        this.bind(new OpLoc4Decoder(), new OpLocHandler())
+        this.bind(new OpLoc5Decoder(), new OpLocHandler())
+        this.bind(new OpObj1Decoder(), new OpObjHandler())
+        this.bind(new OpObj2Decoder(), new OpObjHandler())
+        this.bind(new OpObj3Decoder(), new OpObjHandler())
+        this.bind(new OpObj4Decoder(), new OpObjHandler())
+        this.bind(new OpObj5Decoder(), new OpObjHandler())
+        this.bind(new LocExamineDecoder(), new LocExamineHandler())
```
Comment (Pazaz) [inline src/network/os1/client/prot/game/GameClientRepository.ts]: The examine packets are the 6th option with a special engine-case, really called OPLOC6 OPNPC6 OPOBJ6

**Comment by zeruth** [comment]
Those issues are resolved as well. sorry for not seeing the decoder side of cleanup. I get what you meant now.
Overall the PR is looking much cleaner and does the same job, thanks again!

**Comment by zeruth** [comment]
Did some more cleanup and added a bonus of doing the login packet and fixing the music tab :) I won't be adding to this PR minus any more feedback

**Pull request merged**